nixos/modules/programs/benchexec.nix

```diff
@@ -0,0 +1,59 @@
+{ lib
+, pkgs
+, config
+, options
+, ...
+}:
+let
+  cfg = config.programs.benchexec;
+  opt = options.programs.benchexec;
+in
+{
+  options.programs.benchexec = {
+    enable = lib.options.mkEnableOption "BenchExec";
+    package = lib.options.mkPackageOption pkgs "benchexec" { };
+
+    users = lib.options.mkOption {
+      type = lib.types.listOf lib.types.str;
+      description = lib.mdDoc ''
+        Users that intend to use BenchExec. Control group delegation will be configured via systemd.
+        For more information, see <https://github.com/sosy-lab/benchexec/blob/3.18/doc/INSTALL.md#setting-up-cgroups>.
+      '';
+      default = [ ];
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+    assertions = map
+      (user: {
+        assertion = config.users.users ? ${user};
+        message = ''
+          The user '${user}' intends to use BenchExec (via `${opt.users}`), but is not configured via `${options.users.users}`.
+        '';
+      })
+      cfg.users;
+
+    environment.systemPackages = [ cfg.package ];
+    security.unprivilegedUsernsClone = true;
+    systemd = {
+      enableUnifiedCgroupHierarchy = true;
+
+      # See <https://github.com/sosy-lab/benchexec/blob/3.18/doc/INSTALL.md#setting-up-cgroups>.
+      services = builtins.listToAttrs
+        (user: {
+          name = "user@${builtins.toString config.users.users.${user}.uid}";
+          value.serviceConfig.Delegate = "yes";
+        })
+        cfg.users;
+    };
```

Comment:
The goal is to set `Delegate=yes` for every affected user's [`user@.service`](https://www.freedesktop.org/software/systemd/man/latest/user@.service.html). This is according to [the installation instructions of BenchExec](https://github.com/sosy-lab/benchexec/blob/3.18/doc/INSTALL.md#setting-up-cgroups) with the exception that they suggest to do this for all users of the system. The [maintainer commented that it's a good idea to do it per user if possible](https://github.com/sosy-lab/benchexec/pull/791#issuecomment-1772779159).

Now, there might be better ways to achieve this. Happy to learn.

Reply:
Right, so if I understood this correctly, it has do to with the security implications of letting any user access `CAP_SYS_RAWIO`?

I guess there's a tough choice to make between choosing users individually while lacking mutable user support and generating individual user specific services, versus saying it's a global thing with easier config and to ask people please not to enable it if they are worried about abuse.

Alternatively, it might be possible to implement both, and switch between them using an option. I'm fine with any of these options to be honest, so I think I'll just let this be for now and wait for more opinions.

Reply:
No, it's orthogonal to `CAP_SYS_RAWIO`. `CAP_SYS_RAWIO` is only required to open MSR devices.

This is required because BenchExec sets up control groups for the programs that it benchmarks. And in order for it to be able to do so, delegation of control groups must be enabled.

Reply:
Another, somewhat jank way of solving this might be to add some sort of preExec script to a global user service, check that you are part of a certain group, or else just shut down the unit. I'm not sure if it would be possible to hack your way around using systemd overrides.

Reply:
Sorry, I don't understand. The `preExec` script would have to reconfigure the `user@.service`, which it can't do?!

Reply:
No, more like the user@.service (which should not be modifiable by the user, because it exists in /etc/systemd or the nix store or something), will have a PreExec that will check which user runs the program. If the user does not exist in the correct group, or for any other reasons don't have correct permissions to run the service, PreExec should shut down or intentionally crash such that Exec never gets run.

I haven't ever done such a thing before though, and there are some overriding features in systemd which might make that idea worthless (if the user could just remove the PreExec). I have no clue whether it's even something worth looking into.

Reply:
I thought about this for a bit, and I believe a better solution would be a drop-in override:

```nix
      services = builtins.listToAttrs
        (user: {
          name = "user@${builtins.toString config.users.users.${user}.uid}";
          value = {
            serviceConfig.Delegate = "yes";
            overrideStrategy = "asDropin";
          };
        })
        cfg.users;
```

Reply:
Please check the changes and reply. Otherwise, since this has been stalling for 6 months, I'll just mark it as resolved to unblock the PR.